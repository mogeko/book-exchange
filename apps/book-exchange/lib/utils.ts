import { clsx, type ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function tap<T>(fn: (x: T) => void) {
  return (x: T) => (fn(x), x);
}

export function range(to: number): number[];
export function range(from: number, to: number): number[];
export function range(from: number, to?: number) {
  return Array.from({ length: to ? to - from : from }, (_, i) => i + from);
}
