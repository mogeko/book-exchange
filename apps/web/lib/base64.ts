export { decode, encode } from "js-base64";
