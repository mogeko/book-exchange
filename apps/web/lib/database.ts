import "server-only";

import { PrismaClient } from "@mogeko/bookworm-db";

declare global {
  var prisma: PrismaClient;
}

// Make sure that PrismaClient is instantiated only once.
// See: https://github.com/prisma/prisma/issues/1983#issuecomment-620621213
export const prisma = (() => {
  if (process.env.NODE_ENV === "production") {
    return new PrismaClient();
  } else {
    if (!global.prisma) {
      global.prisma = new PrismaClient();
    }

    return global.prisma;
  }
})();

export * from "@mogeko/bookworm-db";
