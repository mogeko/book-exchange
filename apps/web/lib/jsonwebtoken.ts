import { jwtVerify, SignJWT, type JWTPayload } from "jose";

import { secret } from "@/config/secret";

export async function sign(payland: JWTPayload, { expiresIn, alg }: Opts = {}) {
  return await new SignJWT(payland)
    .setProtectedHeader({ alg: alg ?? "HS256" })
    .setExpirationTime(expiresIn ?? "7d")
    .sign(new TextEncoder().encode(secret.jwt));
}

export async function verify(token: string, { alg }: Pick<Opts, "alg"> = {}) {
  const key = new TextEncoder().encode(secret.jwt);
  const { payload } = await jwtVerify(token, key, {
    algorithms: [alg ?? "HS256"],
  });

  return payload;
}

type Opts = {
  expiresIn?: string | number;
  alg?: string;
};
