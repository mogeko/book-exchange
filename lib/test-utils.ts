import { render, renderHook } from "@testing-library/react";
import type { RenderOptions, RenderHookOptions } from "@testing-library/react";
import WrapProvider from "@/providers/providers";
import { SWRConfig } from "swr";
import React from "react";

function renderWithProvider(ui: React.ReactElement, options?: RenderOptions) {
  return render(ui, { wrapper: WrapProviderForTest, ...options });
}

function renderHookWithProvider<Result, Props>(
  render: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props>
) {
  return renderHook(render, { wrapper: WrapProviderForTest, ...options });
}

const WrapProviderForTest: React.FC<{
  children: React.ReactNode;
}> = ({ children }) => {
  return React.createElement(
    WrapProvider,
    null,
    React.createElement(
      SWRConfig,
      { value: { provider: () => new Map() } },
      children
    )
  );
};

export * from "@testing-library/react";
export { renderHookWithProvider as renderHook };
export { renderWithProvider as render };
