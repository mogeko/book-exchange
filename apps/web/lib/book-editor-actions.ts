"use server";

import { revalidatePath } from "next/cache";

import { prisma } from "@/lib/database";
import type { DataTypes } from "@/components/book-editor";

export async function updateBook(data: DataTypes) {
  try {
    const { id: bid } = await prisma.book.upsert({
      where: { title: data.title },
      update: {
        title: data.title,
        subtitle: data.subtitle,
        description: data.description,
        cover: data.cover,
        authors: {
          connectOrCreate: data?.authors?.map((author) => ({
            where: { name: author.value },
            create: { name: author.value, avatar: "" },
          })),
        },
        translators: {
          connectOrCreate: data?.translators?.map((translator) => ({
            where: { name: translator.value },
            create: { name: translator.value, avatar: "" },
          })),
        },
      },
      create: {
        title: data.title,
        subtitle: data.subtitle,
        description: data.description,
        cover: data.cover,
        authors: {
          connectOrCreate: data?.authors?.map((author) => ({
            where: { name: author.value },
            create: { name: author.value, avatar: "" },
          })),
        },
        translators: {
          connectOrCreate: data?.translators?.map((translator) => ({
            where: { name: translator.value },
            create: { name: translator.value, avatar: "" },
          })),
        },
      },
    });

    return revalidatePath(`/book/${bid}`), {};
  } catch (error: any) {
    return { error: error.message };
  }
}
