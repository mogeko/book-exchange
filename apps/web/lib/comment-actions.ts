"use server";

import { revalidatePath } from "next/cache";

import { prisma, type Comment } from "@/lib/database";
import type { VoteState } from "@/components/comment-context";

export async function removeComment(cid: number) {
  try {
    const comment = await prisma.comment.delete({ where: { id: cid } });

    return revalidatePathHelper(comment);
  } catch (error: any) {
    return { error: error.message };
  }
}

export async function likeDislike(state: VoteState, uid: number, cid: number) {
  console.log(state, cid, uid);
  try {
    if (state) {
      const { comment } = await prisma.voter.upsert({
        where: { voterId_commentId: { commentId: cid, voterId: uid } },
        update: { vote: state },
        create: {
          voter: { connect: { id: uid } },
          comment: { connect: { id: cid } },
          vote: state,
        },
        include: { comment: true },
      });

      return revalidatePathHelper(comment);
    } else {
      const { comment } = await prisma.voter.delete({
        where: { voterId_commentId: { commentId: cid, voterId: uid } },
        include: { comment: true },
      });

      return revalidatePathHelper(comment);
    }
  } catch (error: any) {
    return { error: error.message };
  }
}

const revalidatePathHelper = (comment: Comment) => {
  if (comment.authorId) {
    revalidatePath(`/auther/${comment.authorId}`);
  } else if (comment.publisherId) {
    revalidatePath(`/publisher/${comment.publisherId}`);
  } else if (comment.seriesId) {
    revalidatePath(`/series/${comment.seriesId}`);
  } else if (comment.userId) {
    revalidatePath(`/user/${comment.userId}`);
  }

  return {};
};
