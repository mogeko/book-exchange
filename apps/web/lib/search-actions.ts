"use server";

import { prisma } from "@/lib/database";

export async function search(query: string) {
  return await prisma.book.findMany({
    where: {
      OR: [
        { title: { contains: query } },
        { authors: { some: { name: { contains: query } } } },
        { isbn: { contains: query } },
      ],
    },
    select: { id: true, title: true },
    orderBy: {
      owners: { _count: "desc" },
    },
    take: 10,
    skip: 0,
  });
}
