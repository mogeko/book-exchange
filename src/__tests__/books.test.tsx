import "@testing-library/jest-dom";
import { render, screen, waitFor } from "@/lib/test-utils";
import mockRouter from "next-router-mock";
import BookPage, { BookView } from "@/pages/books/[id]";

describe("Books Page", () => {
  beforeEach(() => {
    mockRouter.setCurrentUrl("/books/bk1000");
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("renders a Books Page", () => {
    const { container } = render(<BookPage />);

    expect(container).toMatchSnapshot();
  });
});

describe("BookView", () => {
  it("render a BookView", async () => {
    const { container } = render(<BookView id="bk1000" />);

    await waitFor(() => {
      expect(screen.getAllByRole("heading")).toHaveLength(3);
      expect(
        screen.getByRole("heading", { name: "About this book" })
      ).toBeInTheDocument();
      expect(
        screen.getByRole("heading", { name: "Popular Highlights in this book" })
      ).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });
});
