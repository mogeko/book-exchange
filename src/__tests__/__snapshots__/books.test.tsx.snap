// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookView render a BookView 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <h1
      class="text-2xl w-full py-2"
    >
      text
    </h1>
    <div
      class="my-4 flex w-full gap-3"
    >
      <figure
        aria-label="Cover Image"
        class="flex items-center justify-center"
      >
        <div
          class="relative w-32 h-48"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="text"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </div>
      </figure>
      <div
        class="flex flex-col"
      >
        <div
          class="inline-flex gap-1 text-base"
        >
          <span
            class="capitalize"
          >
            author
            :
          </span>
          <span>
            Gardner
          </span>
        </div>
        <div
          class="inline-flex gap-1 text-base"
        >
          <span
            class="capitalize"
          >
            publisher
            :
          </span>
          <span>
            Mante Inc
          </span>
        </div>
        <div
          class="inline-flex gap-1 text-base"
        >
          <span
            class="capitalize"
          >
            subtitle
            :
          </span>
          <span>
            Officiis adipisci rerum dolorum cum et totam eveniet culpa qui.
          </span>
        </div>
        <div
          class="inline-flex gap-1 text-base"
        >
          <span
            class="capitalize"
          >
            language
            :
          </span>
          <span>
            Russian
          </span>
        </div>
        <div
          class="inline-flex gap-1 text-base"
        >
          <span
            class="capitalize"
          >
            Publication Date:
          </span>
          <span>
            2002-04-24
          </span>
        </div>
        <div
          class="inline-flex gap-1 text-base"
        >
          <span
            class="uppercase"
          >
            isbn
            :
          </span>
          <span>
            978-1276711770
          </span>
        </div>
        <div
          class="inline-flex gap-1 text-base"
        >
          <span
            class="capitalize"
          >
            hardcover
            :
          </span>
          <span>
            779
             pages
          </span>
        </div>
      </div>
    </div>
    <div
      class="flex flex-col items-start my-2"
    >
      <h2
        class="text-lg capitalize w-full my-2"
      >
        About this book
      </h2>
      <div
        class="flex flex-col items-start"
      >
        <p
          class="indent-8"
        >
          Atque consequatur amet iure sequi voluptas dolorem ipsa. Non quam voluptatibus deserunt omnis. Similique cum aliquid est sunt odio nemo ut inventore est. Ut magnam ea dolores ipsum ea recusandae porro et facere. Quae fuga dolor ipsa dolor fugit unde sit. Sint sint aspernatur ullam a est repellendus. Odio maxime id est dolor nobis. Officiis voluptatem veniam maiores aut laborum perferendis. Quo minus quis consequatur natus nihil dicta et. Placeat et aut earum harum a.
        </p>
        <button
          class="btn btn-sm btn-link"
        >
          <svg
            class="w-5 h-5"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 320 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
            />
          </svg>
           Read more
        </button>
      </div>
    </div>
    <div
      class="flex flex-col items-start my-2"
    >
      <h2
        class="text-lg capitalize w-full my-2"
      >
        Popular Highlights in this book
      </h2>
      <div
        class="flex flex-col items-start"
      >
        <p
          class="indent-8"
        >
          Exercitationem non quia ratione itaque molestias qui libero aspernatur. Corporis in repudiandae aut ipsam soluta. Vel vel facere autem modi quos qui eaque. Ea quibusdam totam dolor officiis et aut est corporis. Voluptas est non fugit occaecati numquam dolor fugit ipsa consequatur. Repellat laboriosam nulla eveniet voluptatem fugiat tempore veritatis vitae cumque. Voluptatum est tempora itaque. Nesciunt unde doloribus est est sed et pariatur. Aut vero impedit quidem pariatur non sequi voluptas. Illo nemo voluptatibus deserunt amet dolorum soluta culpa itaque esse.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Books Page renders a Books Page 1`] = `
<div>
  <div
    class="drawer"
  >
    <input
      aria-hidden="true"
      class="drawer-toggle"
      id="mobile-menu"
      type="checkbox"
    />
    <div
      class="drawer-content"
    >
      <div
        class="flex flex-col min-h-screen min-w-full w-full"
      >
        <div
          class="min-w-full"
        >
          <div
            class="hidden h-7 bg-black justify-between sm:flex"
            data-theme="dracula"
          >
            <nav
              class="navbar min-h-0 flex-"
            >
              <a
                class="btn btn-link btn-xs gap-1 text-gray-400 hover:text-primary focus:text-primary-focus"
                href="/#"
              >
                My library
              </a>
              <a
                class="btn btn-link btn-xs gap-1 text-gray-400 hover:text-primary focus:text-primary-focus"
                href="/#"
              >
                Ranking
              </a>
              <a
                class="btn btn-link btn-xs gap-1 text-gray-400 hover:text-primary focus:text-primary-focus"
                href="/#"
              >
                Quote
              </a>
              <a
                class="btn btn-link btn-xs text-gray-400 hover:text-primary focus:text-primary-focus"
                href="/random"
              >
                I'm Feeling Lucky!
              </a>
            </nav>
            <div
              class="justify-end"
            >
              <div
                class="dropdown dropdown-end"
              >
                <label
                  class="hidden sm:block normal-case"
                  tabindex="0"
                >
                  <div
                    class="animate-pulse w-20 mr-4 h-2 bg-slate-700 rounded"
                  />
                </label>
                <label
                  class="btn btn-ghost btn-circle avatar sm:hidden"
                  tabindex="0"
                >
                  <div
                    class="w-6 rounded-full"
                  >
                    <svg
                      class="animate-spin w-6 h-6"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 16 16"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M13.917 7A6.002 6.002 0 0 0 2.083 7H1.071a7.002 7.002 0 0 1 13.858 0h-1.012z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </div>
                </label>
                <ul
                  aria-label="User Menu"
                  class="menu menu-compact dropdown-content shadow bg-base-300 rounded-box w-52 mt-3 p-2 sm:mt-4 sm:p-0"
                  role="menu"
                  tabindex="0"
                >
                  <li
                    aria-label="User Menu Item"
                    role="menuitem"
                  >
                    <a
                      class="justify-between"
                      href="/#"
                    >
                      <span>
                        Profile
                      </span>
                    </a>
                  </li>
                  <li
                    aria-label="User Menu Item"
                    role="menuitem"
                  >
                    <a
                      class="justify-between"
                      href="/#"
                    >
                      <span>
                        Settings
                      </span>
                    </a>
                  </li>
                  <li
                    aria-label="User Menu Item"
                    role="menuitem"
                  >
                    <button>
                      Sign out
                    </button>
                  </li>
                </ul>
              </div>
            </div>
          </div>
          <div
            class="bg-inherit p-2 md:p-0 md:bg-base-300"
          >
            <header
              class="navbar p-0 m-auto rounded-md max-w-screen-lg bg-base-300 md:rounded-none"
            >
              <div
                class="navbar-start sm:hidden"
              >
                <label
                  aria-label="Drawer Button"
                  class="btn btn-ghost btn-circle"
                  for="mobile-menu"
                  role="button"
                >
                  <svg
                    aria-hidden="true"
                    class="h-6 w-6"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 20 20"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M3 5a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 10a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 15a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </label>
              </div>
              <div
                class="navbar-center sm:navbar-start"
              >
                <div
                  class="flex justify-start"
                >
                  <a
                    class="btn gap-1 btn-ghost normal-case"
                    href="/"
                  >
                    <span
                      class="sr-only"
                    >
                      Logo
                    </span>
                    <span
                      style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                    >
                      <span
                        style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                      >
                        <img
                          alt=""
                          aria-hidden="true"
                          src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2724%27%20height=%2724%27/%3e"
                          style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                        />
                      </span>
                      <img
                        alt="logo"
                        data-nimg="intrinsic"
                        decoding="async"
                        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                      />
                      <noscript />
                    </span>
                    <h1
                      class="text-xl font-bold text-base-content"
                    >
                      Bookworm
                    </h1>
                  </a>
                </div>
              </div>
              <div
                class="navbar-end"
              >
                <div
                  class="hidden md:block relative"
                >
                  <svg
                    class="h-6 w-6 absolute top-3 left-3 text-gray-400"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 20 20"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M8 4a4 4 0 100 8 4 4 0 000-8zM2 8a6 6 0 1110.89 3.476l4.817 4.817a1 1 0 01-1.414 1.414l-4.816-4.816A6 6 0 012 8z"
                      fill-rule="evenodd"
                    />
                  </svg>
                  <div
                    class="form-control"
                  >
                    <input
                      aria-label="Search Bar"
                      class="input bg-base-300 text-xl w-96 pl-10"
                      placeholder="Search..."
                      role="search"
                      type="text"
                    />
                  </div>
                </div>
                <a
                  class="btn btn-ghost btn-circle md:hidden"
                  href="/#"
                  role="button"
                >
                  <svg
                    aria-hidden="true"
                    class="h-6 w-6"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 20 20"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M8 4a4 4 0 100 8 4 4 0 000-8zM2 8a6 6 0 1110.89 3.476l4.817 4.817a1 1 0 01-1.414 1.414l-4.816-4.816A6 6 0 012 8z"
                      fill-rule="evenodd"
                    />
                  </svg>
                  <span
                    class="sr-only"
                  >
                    Search
                  </span>
                </a>
                <div
                  class="sm:hidden"
                >
                  <div
                    class="dropdown dropdown-end"
                  >
                    <label
                      class="hidden sm:block normal-case"
                      tabindex="0"
                    >
                      <div
                        class="animate-pulse w-20 mr-4 h-2 bg-slate-700 rounded"
                      />
                    </label>
                    <label
                      class="btn btn-ghost btn-circle avatar sm:hidden"
                      tabindex="0"
                    >
                      <div
                        class="w-6 rounded-full"
                      >
                        <svg
                          class="animate-spin w-6 h-6"
                          fill="currentColor"
                          height="1em"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 16 16"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            clip-rule="evenodd"
                            d="M13.917 7A6.002 6.002 0 0 0 2.083 7H1.071a7.002 7.002 0 0 1 13.858 0h-1.012z"
                            fill-rule="evenodd"
                          />
                        </svg>
                      </div>
                    </label>
                    <ul
                      aria-label="User Menu"
                      class="menu menu-compact dropdown-content shadow bg-base-300 rounded-box w-52 mt-3 p-2 sm:mt-4 sm:p-0"
                      role="menu"
                      tabindex="0"
                    >
                      <li
                        aria-label="User Menu Item"
                        role="menuitem"
                      >
                        <a
                          class="justify-between"
                          href="/#"
                        >
                          <span>
                            Profile
                          </span>
                        </a>
                      </li>
                      <li
                        aria-label="User Menu Item"
                        role="menuitem"
                      >
                        <a
                          class="justify-between"
                          href="/#"
                        >
                          <span>
                            Settings
                          </span>
                        </a>
                      </li>
                      <li
                        aria-label="User Menu Item"
                        role="menuitem"
                      >
                        <button>
                          Sign out
                        </button>
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
            </header>
          </div>
        </div>
        <main
          class="flex-1 max-w-screen-lg w-full mx-auto p-6 md:p-4"
        >
          <div
            class="animate-pulse"
          >
            <div
              class="flex flex-col"
            >
              <div
                class="h-6 w-40 my-1 bg-slate-700 rounded"
              />
              <div
                class="my-4 flex w-full gap-3"
              >
                <div
                  class="relative w-32 h-48 bg-slate-700 rounded"
                />
                <div
                  class="flex flex-col gap-3"
                >
                  <div
                    class="inline-flex gap-2"
                  >
                    <div
                      class="w-14 h-2 bg-slate-700 rounded"
                    />
                    <div
                      class="w-32 h-2 bg-slate-700 rounded"
                    />
                  </div>
                  <div
                    class="inline-flex gap-2"
                  >
                    <div
                      class="w-14 h-2 bg-slate-700 rounded"
                    />
                    <div
                      class="w-32 h-2 bg-slate-700 rounded"
                    />
                  </div>
                  <div
                    class="inline-flex gap-2"
                  >
                    <div
                      class="w-14 h-2 bg-slate-700 rounded"
                    />
                    <div
                      class="w-32 h-2 bg-slate-700 rounded"
                    />
                  </div>
                  <div
                    class="inline-flex gap-2"
                  >
                    <div
                      class="w-14 h-2 bg-slate-700 rounded"
                    />
                    <div
                      class="w-32 h-2 bg-slate-700 rounded"
                    />
                  </div>
                  <div
                    class="inline-flex gap-2"
                  >
                    <div
                      class="w-14 h-2 bg-slate-700 rounded"
                    />
                    <div
                      class="w-32 h-2 bg-slate-700 rounded"
                    />
                  </div>
                  <div
                    class="inline-flex gap-2"
                  >
                    <div
                      class="w-14 h-2 bg-slate-700 rounded"
                    />
                    <div
                      class="w-32 h-2 bg-slate-700 rounded"
                    />
                  </div>
                  <div
                    class="inline-flex gap-2"
                  >
                    <div
                      class="w-14 h-2 bg-slate-700 rounded"
                    />
                    <div
                      class="w-32 h-2 bg-slate-700 rounded"
                    />
                  </div>
                  <div
                    class="inline-flex gap-2"
                  >
                    <div
                      class="w-14 h-2 bg-slate-700 rounded"
                    />
                    <div
                      class="w-32 h-2 bg-slate-700 rounded"
                    />
                  </div>
                  <div
                    class="inline-flex gap-2"
                  >
                    <div
                      class="w-14 h-2 bg-slate-700 rounded"
                    />
                    <div
                      class="w-32 h-2 bg-slate-700 rounded"
                    />
                  </div>
                  <div
                    class="inline-flex gap-2"
                  >
                    <div
                      class="w-14 h-2 bg-slate-700 rounded"
                    />
                    <div
                      class="w-32 h-2 bg-slate-700 rounded"
                    />
                  </div>
                </div>
              </div>
              <div
                class="flex flex-col items-start my-2"
              >
                <div
                  class="h-4 w-44 mb-4 bg-slate-700 rounded"
                />
                <div
                  class="inline-flex w-full mb-3"
                >
                  <div
                    class="w-8"
                  />
                  <div
                    class="w-full h-2 bg-slate-700 rounded"
                  />
                </div>
                <div
                  class="w-full mb-3 h-2 bg-slate-700 rounded"
                />
                <div
                  class="w-full mb-3 h-2 bg-slate-700 rounded"
                />
                <div
                  class="w-full mb-3 h-2 bg-slate-700 rounded"
                />
                <div
                  class="w-64 h-2 bg-slate-700 rounded"
                />
              </div>
              <div
                class="flex flex-col items-start my-2"
              >
                <div
                  class="h-4 w-44 mb-4 bg-slate-700 rounded"
                />
                <div
                  class="inline-flex w-full mb-3"
                >
                  <div
                    class="w-8"
                  />
                  <div
                    class="w-full h-2 bg-slate-700 rounded"
                  />
                </div>
                <div
                  class="w-full mb-3 h-2 bg-slate-700 rounded"
                />
                <div
                  class="w-full mb-3 h-2 bg-slate-700 rounded"
                />
                <div
                  class="w-full mb-3 h-2 bg-slate-700 rounded"
                />
                <div
                  class="w-64 h-2 bg-slate-700 rounded"
                />
              </div>
            </div>
          </div>
        </main>
        <footer
          aria-label="Footer"
          class="footer footer-center p-4 bg-base-300 text-base-content"
        >
          <div>
            <p
              aria-label="copyright"
              role="presentation"
            >
              Copyright © 
              2022
              , All right reserved
               by Zheng Junyi
              .
            </p>
          </div>
        </footer>
      </div>
    </div>
    <div
      class="drawer-side"
    >
      <label
        class="drawer-overlay"
        for="mobile-menu"
      />
      <ul
        aria-label="Drawer Menu"
        class="menu p-4 overflow-y-auto w-80 bg-base-200 text-base-content"
        role="menu"
      >
        <li
          role="menuitem"
        >
          <a
            class="justify-between"
            href="/#"
          >
            <span>
              My library
            </span>
          </a>
        </li>
        <li
          role="menuitem"
        >
          <a
            class="justify-between"
            href="/#"
          >
            <span>
              Ranking
            </span>
          </a>
        </li>
        <li
          role="menuitem"
        >
          <a
            class="justify-between"
            href="/#"
          >
            <span>
              Quote
            </span>
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
